src/mixins/itext_behavior.mixin.js

```diff
@@ -412,6 +418,252 @@
       }
     },
 
+    /**
+     * Override to customize the drag image
+     * https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/setDragImage
+     * @param {DragEvent} e
+     * @param {object} data
+     * @param {number} data.selectionStart
+     * @param {number} data.selectionEnd
+     * @param {boolean} data.trailingSpace
+     * @param {string} data.text
+     * @param {string} data.value selected text
+     */
+    setDragImage: function (e, data) {
+      var t = this.calcTransformMatrix();
+      var flipFactor = new fabric.Point(this.flipX ? -1 : 1, this.flipY ? -1 : 1);
+      var boundaries = this._getCursorBoundaries(data.selectionStart);
+      var selectionPosition = new fabric.Point(
+        boundaries.left + boundaries.leftOffset,
+        boundaries.top + boundaries.topOffset
+      ).multiply(flipFactor);
+      var pos = fabric.util.transformPoint(selectionPosition, t);
+      var pointer = this.canvas.getPointer(e);
+      var diff = pointer.subtract(pos);
+      var enableRetinaScaling = this.canvas._isRetinaScaling();
+      var retinaScaling = this.canvas.getRetinaScaling();
+      var bbox = this.getBoundingRect(true);
+      var correction = pos.subtract(new fabric.Point(bbox.left, bbox.top));
+      var offset = correction.add(diff).scalarMultiply(retinaScaling);
+      //  prepare instance for drag image snapshot by making all non selected text invisible
+      var bgc = this.backgroundColor;
+      var styles = fabric.util.object.clone(this.styles, true);
+      delete this.backgroundColor;
+      var styleOverride = {
+        fill: 'transparent',
+        textBackgroundColor: 'transparent'
+      };
+      this.setSelectionStyles(styleOverride, 0, data.selectionStart);
+      this.setSelectionStyles(styleOverride, data.selectionEnd, data.text.length);
+      var dragImage = this.toCanvasElement({ enableRetinaScaling: enableRetinaScaling });
+      this.backgroundColor = bgc;
+      this.styles = styles;
+      //  handle retina scaling
+      if (enableRetinaScaling && retinaScaling > 1) {
+        var c = fabric.util.createCanvasElement();
+        c.width = dragImage.width / retinaScaling;
+        c.height = dragImage.height / retinaScaling;
+        var ctx = c.getContext('2d');
+        ctx.scale(1 / retinaScaling, 1 / retinaScaling);
+        ctx.drawImage(dragImage, 0, 0);
```

Comment:
don't forget to account for viewport transform

Reply:
also, would it not be better to keep the dragImage in the original resolution (with retinascaling) and just resize the drag image's dom element so that it stays in high res?

Reply:
Please elaborate

Reply:
I tested on a high dpi display. The drag image is at a lower dpi than the canvas.

Reply:
I have seen that as well.
Any suggestions?

Reply:
just spent some time looking into this... your code looks correct, I'm really not sure what's causing the dragimage to have a lower res 😐

Reply:
it is not retina scaled because there is probably no css for the drawImage.
indeed here https://github.com/fabricjs/fabric.js/pull/7802/files#diff-af5c45c9018fd92f35ba43f5e25303b5fe3b627d0f81d2d78b328746fe015b4eR468-R473 we are dividing the natural size of the image by the retina factor, making it smaller.
Not sure if the dragged image can targeted by a selector, or can be wrapped in a div and added a class that would retina-scale it.
Is not that much important we can try later

Reply:
If i remember correctly, when handling drag/drop with react.js i remember the dragged element can be anything. I believe you can remove the division by retina scaling and add a div around the canvas that will divide the size by 2 with css and so get the hiDpi effect.